components/engine_rocks/src/ttl_properties.rs

```diff
@@ -0,0 +1,152 @@
+// Copyright 2021 TiKV Project Authors. Licensed under Apache-2.0.
+
+use std::collections::HashMap;
+
+use crate::{RocksEngine, UserProperties};
+use engine_traits::util::get_expire_ts;
+use engine_traits::{
+    DecodeProperties, Range, Result, TtlProperties, TtlPropertiesExt, TableProperties,
+    TablePropertiesCollection, TablePropertiesExt,
+};
+use rocksdb::{DBEntryType, TablePropertiesCollector, TablePropertiesCollectorFactory};
+
+const PROP_MAX_EXPIRE_TS: &str = "tikv.max_expire_ts";
+const PROP_MIN_EXPIRE_TS: &str = "tikv.min_expire_ts";
+
+pub struct RocksTtlProperties;
+
+impl RocksTtlProperties {
+    pub fn encode(ttl_props: &TtlProperties) -> UserProperties {
+        let mut props = UserProperties::new();
+        props.encode_u64(PROP_MAX_EXPIRE_TS, ttl_props.max_expire_ts);
+        props.encode_u64(PROP_MIN_EXPIRE_TS, ttl_props.min_expire_ts);
+        props
+    }
+
+    pub fn decode<T: DecodeProperties>(props: &T) -> Result<TtlProperties> {
+        let mut res = TtlProperties::default();
+        res.max_expire_ts = props.decode_u64(PROP_MAX_EXPIRE_TS)?;
+        res.min_expire_ts = props.decode_u64(PROP_MIN_EXPIRE_TS)?;
+        Ok(res)
+    }
+}
+
+impl TtlPropertiesExt for RocksEngine {
+    fn get_range_ttl_properties_cf(
+        &self,
+        cf: &str,
+        start_key: &[u8],
+        end_key: &[u8],
+    ) -> Result<Vec<(String, TtlProperties)>> {
+        let range = Range::new(start_key, end_key);
+        let collection = self.get_properties_of_tables_in_range(cf, &[range])?;
+        if collection.is_empty() {
+            return Ok(vec![]);
+        }
+
+        let mut res = Vec::new();
+        for (file_name, v) in collection.iter() {
+            let prop = match RocksTtlProperties::decode(&v.user_collected_properties()) {
+                Ok(v) => v,
+                Err(_) => continue,
+            };
+            res.push((file_name.to_string(), prop));
+        }
+        Ok(res)
+    }
+}
+
+#[derive(Default)]
+/// Can only be used for default CF.
+pub struct TtlPropertiesCollector {
+    prop: TtlProperties,
+}
+
+impl TablePropertiesCollector for TtlPropertiesCollector {
+    fn add(&mut self, key: &[u8], value: &[u8], entry_type: DBEntryType, _: u64, _: u64) {
+        if entry_type != DBEntryType::Put {
+            return;
+        }
+        // only consider data keys
+        if !key.starts_with(keys::DATA_PREFIX_KEY) {
+            return;
+        }
+
+        let expire_ts = get_expire_ts(&value).unwrap_or_else(|_| {
+            error!("unexpected ttl key:{:?}, value:{:?}", key, value);
```

Comment:
This probably violates the log redaction requirements.

Reply:
remove the log, and the error will be recorded by metrics in later PR